File: Scala/src/main/scala/iterator/PancakeHouseMenu.scala
package iterator

import java.util

class PancakeHouseMenu extends Menu {
  override val name: String = "아침 메뉴"

  private val menu = new util.ArrayList[MenuItem]
  addItem(MenuItem("K&B 팬케이크 세트", "스크램블 에그와 토스트가 곁들여진 팬케이크", vegetarian = true, 2.99))
  addItem(MenuItem("레귤러 팬케이스 세트", "달걀프라이와 소시지가 곁들여진 팬케이크", vegetarian = false, 2.99))
  addItem(MenuItem("블루베리 팬케이크", "신선한 블루베리와 블루베리 시럽으로 만든 팬케이크", vegetarian = true, 3.49))
  addItem(MenuItem("와플", "취향에 따라 블루베리나 딸기를 얹을 수 있는 와플", vegetarian = true, 3.59))

  override def addItem(menuItem: MenuItem): Unit = menu.add(menuItem)

  def createIterator(): PancakeHouseMenuIterator = new PancakeHouseMenuIterator(menu)

  override def createUtilIterator: util.Iterator[MenuItem] = menu.iterator()
}

File: Kotlin/src/main/kotlin/chapter9_iterator/PancakeHouseMenu.kt
package chapter9_iterator

import java.math.BigDecimal

class PancakeHouseMenu {
    var menuItems: MutableList<MenuItem> = mutableListOf();

    init {
        addItem(
            name = "K&B 팬케이크 세트",
            description = "스크램블 에그와 토스트가 곁들여진 팬케이크",
            vegetarian = true,
            price = BigDecimal(2.99),
        )

        addItem(
            name = "레귤러 팬케이크 세트",
            description = "달걀 프라이와 소시지가 곁들여진 팬케이크",
            vegetarian = false,
            price = BigDecimal(2.99),
        )
    }

    private fun addItem(name: String, description: String, vegetarian: Boolean, price: BigDecimal) {
        menuItems.add(
            MenuItem(name = name, description = description, vegetarian = vegetarian, price = price)
        )
    }

    fun getIterator(): Iterator = PancakeIterator(menuItems)
}